agent/lib/kontena/observable.rb

```diff
@@ -1,75 +1,201 @@
 module Kontena
-  # An Actor that has some value
-  # The value does not yet exist when initialized, it is nil
-  # Once the value is first updated, then other Actors will be able to observe it
-  # When the value later updated, other Actors will also observe those changes
-  module Observable
+  # The observable value is nil when initialized, and Observers will block waiting for it to become ready.
+  # Once the observable is first updated, then Observers will unblock and return/yield the updated value.
+  # When the observable is later updated, then Observers will return/yield the updated value.
+  # If the observable is crashed, then any Observers will also raise.
+  #
+  # TODO: are you allowed to reset an observable after crashing it, allowing observers to restart and re-observe?
```

Comment:
I think it should be possible for an observable actor to do something along the lines of "lightweight crashes":

```
observable.crash FooError.new(...)
observable.reset
# some time later...
observable.update foo
```

And have all observers raise, but allow them to re-observe the same observable again without needing to crash all the associated actors, spawn a new actor with a new observable, and have the observing actors lookup the new observable.

Not sure if that's a good idea, though.

It should work though, because the `Observer` implementation always checks for crashes immediately after each update, so it should be guaranteed to raise. Risk is somehow having the observer only check for `error?` after applying both the `crash` and `reset` updates, which would miss the rror.

Reply:
The main alternative that I have in mind for this is some kind of higher-level layer where the observable actors to update with some kind of "unhealthy" error state, without actually crashing the observable. The observing actors would then enter some kind of error state of their own.

With that in mind, it might be a good idea to just remove the `reset` entirely. I'm not sure if it's actually useful at all.

Reply:
> The observing actors would then enter some kind of unhealthy error state of their own

Isn't that kinda opposite to one of the main actor principles of "let it crash" :) 

I think that kind of state dependencies will get one into trouble really fast. Better to keep things simple and just "let it  🔥 "

Reply:
Yes, actors should be able to crash, and the system should be able to recover itself into a safe state from actor crashes.

But there's also a difference between known and unknown errors. Should the `LogWorker` crash if the websocket disconnects? Can it crash without losing the contents of the `@queue`?

I'm mostly talking about the launcher actors here, that are responsible for the state of some external system. The kind of unhealthy states that I'm talking about here are along the lines of "etcd isn't working because weave was unable to connect to any of its peers", and I don't think it's appropriate for actors to start crashing in that case.

Should the observables start crashing in that case? Probably not, those kind of known errors should be handled as observable `{status: healthy/unhealthy}` states.

Reply:
> But there's also a difference between known and unknown errors

Yes, absolutely. I was also more referring to the un-expected cases. 

> ... known errors should be handled as observable `{status: healthy/unhealthy}` states.

I don't think there's anything in observables that doesn't allow this already? OR are you saying that this kind of "error" update should be done through separate "interface" and not as normal observable update?

Reply:
> I don't think there's anything in observables that doesn't allow this already? OR are you saying that this kind of "error" update should be done through separate "interface" and not as normal observable update?

I think what I'm saying is that we should just remove the `reset` interface entirely... maybe. I can't say for certain before actually validating the design over in#2582

Reply:
Note that there isn't actually any code in the `Observer` to explicitly handle the reset case. The behavior of the `Observable#reset` method is just a side effect of how the `Observer#eval` `ready?` check is implemented.

Reply:
> I think what I'm saying is that we should just remove the reset interface entirely... maybe. I can't say for certain before actually validating the design over in #2582

It was added as an experiment #2011, nothing uses it yet, and based on how it has behaved so far in #2582, I don't expect it to be used.